File: packages/contracts/headless/src/modal.rs
//! Modal machine (Dialog, AlertDialog, Drawer). Mirror of core `modal.ts`.

#[path = "generated/machines/modal.rs"]
mod modal_interface;
pub use modal_interface::*;

fn open() -> (ModalState, Vec<ModalEffect>) {
    (
        ModalState::Open,
        vec![
            ModalEffect::EmitOpenChange { open: true },
            ModalEffect::SaveFocusAndEnter,
            ModalEffect::LockBodyScroll,
        ],
    )
}

fn close(requested: bool) -> (ModalState, Vec<ModalEffect>) {
    let mut effects = Vec::new();

    if requested {
        effects.push(ModalEffect::EmitRequestClose);
    }

    effects.extend([
        ModalEffect::EmitOpenChange { open: false },
        ModalEffect::UnlockBodyScroll,
        ModalEffect::RestoreFocus,
    ]);

    (ModalState::Closed, effects)
}

pub fn modal_transition(
    state: ModalState,
    context: ModalContext,
    event: ModalEvent,
) -> (ModalState, Vec<ModalEffect>) {
    match (state, event) {
        (ModalState::Closed, ModalEvent::Open) => open(),
        (ModalState::Open, ModalEvent::Close) => close(false),
        (ModalState::Open, ModalEvent::RequestClose) => close(true),
        (ModalState::Open, ModalEvent::Escape) if context.dismiss_on_escape => close(true),
        (ModalState::Open, ModalEvent::BackdropClick) if context.dismiss_on_backdrop => close(true),
        _ => (state, vec![]),
    }
}

File: packages/core/src/modal.ts
/**
 * Modal overlay behavior machine (Dialog, AlertDialog via Dialog, Drawer).
 * Contracts: docs/contracts/components/dialog.md, alert-dialog.md, drawer.md,
 * "Behavior Machine" sections.
 *
 * States: closed | open. User-initiated close paths (escape, backdrop, close
 * button) emit `emitRequestClose` before `emitOpenChange(false)`, preserving
 * the onRequestClose → onOpenChange ordering. Open/close side work (focus
 * save + trap entry, focus restore, body scroll lock) are effect intents the
 * adapter executes.
 */

import type { TransitionResult } from "./machine";

export type ModalState = "closed" | "open";

export interface ModalContext {
  dismissOnEscape: boolean;
  dismissOnBackdrop: boolean;
}

export type ModalEvent =
  | { type: "OPEN" }
  | { type: "CLOSE" }
  | { type: "REQUEST_CLOSE" }
  | { type: "ESCAPE" }
  | { type: "BACKDROP_CLICK" };

export type ModalEffect =
  | { type: "emitOpenChange"; open: boolean }
  | { type: "emitRequestClose" }
  | { type: "saveFocusAndEnter" }
  | { type: "restoreFocus" }
  | { type: "lockBodyScroll" }
  | { type: "unlockBodyScroll" };

export type ModalResult = TransitionResult<ModalState, ModalContext, ModalEffect>;

function open(context: ModalContext): ModalResult {
  return {
    state: "open",
    context,
    effects: [
      { type: "emitOpenChange", open: true },
      { type: "saveFocusAndEnter" },
      { type: "lockBodyScroll" },
    ],
  };
}

function close(context: ModalContext, requested: boolean): ModalResult {
  const effects: ModalEffect[] = requested ? [{ type: "emitRequestClose" }] : [];

  return {
    state: "closed",
    context,
    effects: [
      ...effects,
      { type: "emitOpenChange", open: false },
      { type: "unlockBodyScroll" },
      { type: "restoreFocus" },
    ],
  };
}

export function modalTransition(
  state: ModalState,
  context: ModalContext,
  event: ModalEvent,
): ModalResult {
  const stay: ModalResult = { state, context, effects: [] };

  switch (event.type) {
    case "OPEN":
      return state === "closed" ? open(context) : stay;
    case "CLOSE":
      return state === "open" ? close(context, false) : stay;
    case "REQUEST_CLOSE":
      return state === "open" ? close(context, true) : stay;
    case "ESCAPE":
      return state === "open" && context.dismissOnEscape ? close(context, true) : stay;
    case "BACKDROP_CLICK":
      return state === "open" && context.dismissOnBackdrop ? close(context, true) : stay;
  }
}
